### PR TITLE
Plumb in custom vxlan port for Windows nodes

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Add hybrid overlay vxlan communication

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
Depends on https://github.com/openshift/release/pull/10801

File: pkg/clusternetwork/config.go [hunk 1]
```diff
@@ -131,6 +148,24 @@ func getServiceNetworkCIDR(oclient configclient.Interface) (string, error) {
...
+// getVXLANPort gets the VXLAN port to establish tunnel as a string. The return type doesn't matter as we want to pass
+// this argument to a powershell command
+func getVXLANPort(operatorClient operatorv1.OperatorV1Interface) (string, error) {
+	// Get the cluster network object so that we can find the service network
+	networkCR, err := operatorClient.Networks().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrap(err, "error getting cluster network object")
+	}
+	var vxlanPort *uint32
+	if networkCR.Spec.DefaultNetwork.OVNKubernetesConfig != nil &&
```
Comment (VaishnaviHire) [inline pkg/clusternetwork/config.go:160]: Can we use `Validate()` func here ?

File: pkg/clusternetwork/config_test.go [hunk 2]
```diff
@@ -107,3 +113,45 @@ func createFakeClients(networkType string) (configclient.Interface, operatorclie
...
+		wantErr      bool
+	}{
+		{
+			name: "custom vxlan",
+			want: "4800",
+			networkPatch: []byte(`{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"hybridOverlayConfig": ` +
+				`{"hybridClusterNetwork":[{"cidr":"10.132.0.0/14","hostPrefix":23}],"` +
+				`hybridOverlayVXLANPort": 4800}}}}}`),
+			wantErr: false,
+		},
```
Comment (VaishnaviHire) [inline pkg/clusternetwork/config_test.go:131]: Should we also have test with incorrect port value.

File: pkg/clusternetwork/config.go [hunk 3]
```diff
@@ -84,6 +96,11 @@ func (ovn *ovnKubernetes) GetServiceCIDR() (string, error) {
 	return ovn.clusterNetworkConfig.serviceCIDR, nil
 }
 
+// GetVXLANPort gets the vxlan port to be used for vxlan tunnel establishment
+func (ovn *ovnKubernetes) GetVXLANPort() (string, error) {
```
Comment (aravindhp) [inline pkg/clusternetwork/config.go:100]: Why is this function returning an error?

File: pkg/clusternetwork/config_test.go [hunk 4]
```diff
@@ -107,3 +113,45 @@ func createFakeClients(networkType string) (configclient.Interface, operatorclie
 	}
 	return fakeConfigClient, fakeOperatorClient
 }
+
+func Test_getVXLANPort(t *testing.T) {
```
Comment (aravindhp) [inline pkg/clusternetwork/config_test.go:117]: Please add a doc,

File: pkg/clusternetwork/config.go [hunk 5]
```diff
@@ -16,6 +17,7 @@ const ovnKubernetesNetwork = "OVNKubernetes"
 type ClusterNetworkConfig interface {
```
Comment (aravindhp) [inline pkg/clusternetwork/config.go:17]: Side comment: I wish we had just called this `Config`. `clusternetwork.ClusterNetworkConfig` stutters :-(.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 6]
```diff
@@ -91,10 +91,12 @@ type windows struct {
 	signer ssh.Signer
 	// interact is used to connect to and interact with the VM
 	interact connectivity
+	// vxlanPort is the custom vxlan port
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:94]: `VXLAN port`

File: pkg/controller/windowsmachine/windows/windows.go [hunk 7]
```diff
@@ -110,7 +112,8 @@ func New(ipAddress, instanceID, workerIgnitionEndpoint string, signer ssh.Signer
 	return &windows{
 			id:                     instanceID,
 			interact:               conn,
-			workerIgnitionEndpoint: workerIgnitionEndpoint},
+			workerIgnitionEndpoint: workerIgnitionEndpoint,
+			vxlanPort:              vxlanPort},
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:116]: Please move `}` to the next. It is just easier to read.

File: pkg/clusternetwork/config.go (hunk 3 above)
Comment (ravisantoshgudimetla) [reply pkg/clusternetwork/config.go]: Ya, it was a miss on our side earlier, actually these getters won't return an error at all. I just followed the pattern. We should change this perhaps in another PR.

File: pkg/clusternetwork/config.go (hunk 5 above)
Comment (ravisantoshgudimetla) [reply pkg/clusternetwork/config.go:17]: Next PR for sure

File: go.mod [hunk 8]
```diff
@@ -45,10 +47,9 @@ require (
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.5.1
 	golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
-	golang.org/x/net v0.0.0-20200707034311-ab3426394381 // indirect
 	golang.org/x/sys v0.0.0-20200728102440-3e129f6d46b1 // indirect
-	k8s.io/api v0.18.3
-	k8s.io/apimachinery v0.18.3
+	k8s.io/api v0.19.0-rc.2
+	k8s.io/apimachinery v0.19.0-rc.2
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply go.mod:52]: I think we might need it for operator-sdk bump, let's see if we can get to it as part of that. If not, we can re-open

File: go.mod [hunk 9]
```diff
@@ -7,9 +7,11 @@ replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503
 replace (
 	github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.3.2+incompatible // Required by OLM
 	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.38.1-0.20200424145508-7e176fda06cc
+	github.com/go-logr/logr => github.com/go-logr/logr v0.2.1-0.20200730175230-ee2de8da5be6 // Compatibility issues - https://github.com/go-logr/logr/issues/19, bumping to latest operator-sdk might remove this
```
Comment (ravisantoshgudimetla) [inline go.mod:10]: I updated the comment here,

File: cmd/manager/main.go [hunk 10]
```diff
@@ -172,15 +172,8 @@ func main() {
 		os.Exit(1)
```
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply cmd/manager/main.go:172]: Done.

File: test/e2e/clusterinfo/openshift.go [hunk 11]
```diff
@@ -4,16 +4,18 @@ import (
 	"context"
 	"fmt"
 
-	v1 "github.com/openshift/api/config/v1"
+	configv1 "github.com/openshift/api/config/v1"
 	clientset "github.com/openshift/client-go/config/clientset/versioned"
+	operatorv1 "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
```
Comment (aravindhp) [inline test/e2e/clusterinfo/openshift.go:9]: Please use `config` and `operator`

File: test/e2e/network_test.go [hunk 12]
```diff
@@ -456,6 +456,13 @@ func getWinCurlerCommand(winServerIP string) []string {
 func (tc *testContext) createWindowsServerJob(name string, command []string) (*batchv1.Job, error) {
 	windowsNodeSelector := map[string]string{"beta.kubernetes.io/os": "windows"}
 	windowsTolerations := []v1.Toleration{{Key: "os", Value: "Windows", Effect: v1.TaintEffectNoSchedule}}
+	var windowsServerImage string
+	// If we're using a custom VXLANPort we need to use 1909 or else use Windows Server 2019
+	if tc.hasCustomVXLAN {
+		windowsServerImage = "mcr.microsoft.com/powershell:lts-nanoserver-1909"
+	} else {
+		windowsServerImage = "mcr.microsoft.com/powershell:lts-nanoserver-1809"
+	}
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply test/e2e/network_test.go:466]: I don't follow. It does seem like: 
```
    // If we're using a custom VXLANPort we need to use 1909 or else use Windows Server 2019
	if tc.hasCustomVXLAN {
		windowsServerImage = "mcr.microsoft.com/powershell:lts-nanoserver-1909"
	} else {
		windowsServerImage = "mcr.microsoft.com/powershell:lts-nanoserver-1809"
	}
```
is repeated code.

File: test/e2e/providers/aws/aws.go [hunk 13]
```diff
@@ -118,17 +119,23 @@ func (a *awsProvider) getInfraID() (string, error) {
...
 	// This filter will grab all ami's that match the exact name. The '?' indicate any character will match.
 	// The ami's will have the name format: Windows_Server-2019-English-Full-ContainersLatest-2020.01.15
 	// so the question marks will match the date of creation
 	// The image obtained by using windowsAMIFilterValue is compatible with  the test container image -
-	// "mcr.microsoft.com/powershell:lts-nanoserver-1809". If the windowsAMIFilterValue changes,
-	// the test container image also needs to be changed.
-	windowsAMIFilterValue := "Windows_Server-2019-English-Full-ContainersLatest-????.??.??"
+	// "mcr.microsoft.com/powershell:lts-nanoserver-1909" or "mcr.microsoft.com/powershell:lts-nanoserver-1809".
+	// If the windowsAMIFilterValue changes, the test container image also needs to be changed.
+	// if vxlanport is set use 1909 image as it has the custom VXLAN port changes, if not use Windows Server 2019 image
```
Comment (aravindhp) [inline test/e2e/providers/aws/aws.go:133]: s/vxlanport/hasCustomVXLANPort

**Comment by ravisantoshgudimetla** [comment]
> https://github.com/openshift/windows-machine-config-operator/pull/96#discussion_r476837134

Missed this, you guys are right. I made the code changes you suggested

File: go.mod [hunk 14]
```diff
@@ -7,9 +7,11 @@ replace github.com/docker/docker => github.com/moby/moby v0.7.3-0.20190826074503
 replace (
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply go.mod:7]: Please address this.

File: pkg/clusternetwork/config_test.go (hunk 4 above)
Comment (aravindhp) [reply pkg/clusternetwork/config_test.go:118]: Good point. I suggest removing the `_`.

File: pkg/controller/windowsmachine/windowsmachine_controller.go [hunk 15]
```diff
@@ -73,11 +74,19 @@ func newReconciler(mgr manager.Manager, clusterServiceCIDR string) (reconcile.Re
 		return nil, errors.Wrapf(err, "error creating signer using private key: %v", wkl.PrivateKeyPath)
 	}
 
+	serviceCIDR, err := networkConfig.GetServiceCIDR()
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting service CIDR")
+	}
+
+	vxlanPort := networkConfig.VXLANPort()
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windowsmachine_controller.go:82]: Given there is no error being returned, do we need this variable? I suggest you make the call directly below.

File: test/e2e/network_test.go [hunk 16]
```diff
@@ -322,10 +316,23 @@ func (tc *testContext) getPodIP(selector metav1.LabelSelector) (string, error) {
 	return podList.Items[0].Status.PodIP, nil
 }
 
+// getRelevantWindowsServerContainerImage gets the appropriate WindowsServer image based on VXLAN port
+func (tc *testContext) getRelevantWindowsServerContainerImage() string {
```
Comment (aravindhp) [inline test/e2e/network_test.go:320]: `getWindowsServerContainerImage()` will suffice. Using terms like `Relevenat` `Valid` sort of implies we have a use case where we need to get irrelevant or invalid entities.

**Pull request merged**